### ISSUE TITLE
Orchestration concurrency not being enforced

### ISSUE THREAD

**Issue by davidychow87** [Azure/azure-functions-durable-extension#1272]
I have a service bus triggered function, which upon receiving a message, starts a Durable Functions Orchestration Function (using orchestrationClient.StartNewAsync)

This Orchestration function works like so:

1 – Split out work to multiple activity functions (which do processing), which we'll call **ActivityFunction1**
2 – Wait for all of the **ActivityFunction1** instances to complete
3 – Split out work to multiple activity functions, which make Http requests, we'll call these functions **ActivityFunction2**
4 – Do some polling for results, and when the expected results are complete, send a response message.

I’ve tried to limit the orchestration concurrency to 2 (see the host.json below). The issue is that if I send multiple messages (say 5 messages) at the same time, it appears that the durable function will first call the **ActivityFunction1** activity function for all 5 messages, and when those have completed, it will then call the **ActivityFunction2** functions for all 5 at the same time. Ideally, I would like it to finish an existing orchestration before starting a new one (with a concurrency of 2 orchestrations running at a time).

It appears that perhaps Durable functions is prioritizing the **ActivityFunction1** activity function for new orchestrations vs starting the **ActivityFunction2** function for existing orchestrations? Note that for memory/cpu reasons, I’ve set maxConcurrentActivityFunctions to 8.

Here is my host.json:

{
  "version": "2.0",
  "functionTimeout": "04:55:00",
  "extensions": {
    "durableTask": {
      "maxConcurrentActivityFunctions": 8,
      "maxConcurrentOrchestratorFunctions": 2,
      "controlQueueBatchSize":  2
    },
    "serviceBus": {
      "messageHandlerOptions": {
        "maxConcurrentCalls": 2
      }
    }
  },
  "logging": {
    "applicationInsights": {
      "samplingSettings": {
        "isEnabled": false
      }
    },
    "logLevel": {
      "default": "Information"
    }
  }
}

**Comment by sebastianburckhardt** [comment]
Hi @davidychow87 thanks for sharing your situation. This is an interesting observation.

"Unfortunately" I think this is working as designed - though I understand it is not quite what you need. What is happening is that if there are multiple orchestrations issued at the same time, they all call the first activity, which queues those activities for execution. The scheduler is roughly FIFO so it will execute all of those first before any of the second activities, and so on. 

This would not be an issue if `maxConcurrentOrchestratorFunctions` would prevent new orchestrations from starting before old ones complete. However, that is not how it works. The `maxConcurrentOrchestratorFunctions` does not limit the number of orchestrations that are in flight, it only limits the number of orchestration steps that are being processed at any one moment on any one compute node. 

I can't think of an easy way to effectively limit the number of orchestrations that get started (which would help older orchestrations to finish before starting new ones). Perhaps we need to add some form of support for that.

**Comment by amadard** [comment]
@sebastianburckhardt I would very much like a "singleton" pattern that supports limiting the number of concurrently running orchestrations.  The [singleton pattern documentation](https://docs.microsoft.com/en-us/azure/azure-functions/durable/durable-functions-singletons?tabs=csharp) implies that we can use the instance id, but that only restricts the number **ever** started.  

The current singleton pattern has only one instance ever created, rather than ensuring only one instance is running at a time.  There is a need to ensure only X instance(s) is/are running concurrently, and there isn't a good way in the functions to do that as far as I know.  And that is what @davidychow87 is asking for, if I am understanding correctly.

**Comment by sebastianburckhardt** [comment]
My best idea right now for how to get this working (without special runtime support) is to use an entity as an "OrchestrationLimiter". Basically, instead of starting the orchestration directly, you ask the limiter to do it - but it will only do it if the limit is not reached yet, otherwise it queues the orchestration instead of starting it. The other part is that orchestrations have to tell the limiter when they are finished. 

Here is some sample code for it:

```csharp
public interface IOrchestrationLimiter
{
    void StartNewOrchestration((string functionName, object input, string instanceId) input);
    void FinishedOne();
}

[JsonObject(MemberSerialization.OptIn)]
public class OrchestrationLimiter : IOrchestrationLimiter
{
    private const int limit = 1;  // or any other number

    [JsonProperty()]
    public int numberCurrentlyRunning = 0;

    [JsonProperty()]
    public Queue<(string, object, string)> Queue { get; set; } = new Queue<(string, object, string)>();

    public void StartNewOrchestration((string functionName, object input, string instanceId) input)
    {
        if (numberCurrentlyRunning < limit)
        {
            Entity.Current.StartNewOrchestration(input.functionName, input.input, input.instanceId);
            numberCurrentlyRunning++;
        }
        else
        {
            Queue.Enqueue(input);
        }
    }

    public void FinishedOne()
    {
        numberCurrentlyRunning--;

        if (numberCurrentlyRunning < limit && Queue.Count > 0)
        {
            var (functionName, input, instanceId) = Queue.Dequeue();
            Entity.Current.StartNewOrchestration(functionName, input, instanceId);
            numberCurrentlyRunning++;
        }
    }

    [FunctionName(nameof(OrchestrationLimiter))]
    public static Task Run([EntityTrigger] IDurableEntityContext ctx) => ctx.DispatchAsync<OrchestrationLimiter>();
}

public static class DemoFunctions
{
    [FunctionName("ExampleOrchestration")]
    public static async Task ExampleOrchestration([OrchestrationTrigger] IDurableOrchestrationContext ctx)
    {
        try
        {
            await ctx.CreateTimer(DateTime.UtcNow + TimeSpan.FromSeconds(10), CancellationToken.None);
        }
        finally // want to make sure this always happens when orchestration finishes, with or without exception
        {
            ctx.CreateEntityProxy<IOrchestrationLimiter>("globalLimiter").FinishedOne();
        }
    }

    [FunctionName("Start")]
    public static async Task<IActionResult> Start(
        [HttpTrigger(AuthorizationLevel.Anonymous, "post")] HttpRequest req,
        [DurableClient] IDurableClient client)
    {
        string instanceId = Guid.NewGuid().ToString();
        await client.SignalEntityAsync<IOrchestrationLimiter>("globalLimiter", p => p.StartNewOrchestration(("ExampleOrchestration", "", instanceId)));
        return await client.WaitForCompletionOrCreateCheckStatusResponseAsync(req, instanceId, TimeSpan.FromSeconds(60));
    }    
}
```

**Comment by amadard** [comment]
@sebastianburckhardt That is a really neat way of fixing the issue.  My coworker did have the thought that we could look up the status of all running instances and filter by name to find out the count, but I like using the durable entity much better.  I am going to review this for our use cases.

Thank you!

**Comment by davidychow87** [comment]
@amadard Yep, that is what I mean

@sebastianburckhardt Thanks for the code example, I will try that out. I have also been told that perhaps suborchestrations might work (https://docs.microsoft.com/en-us/azure/azure-functions/durable/durable-functions-sub-orchestrations?tabs=csharp). If I wrap the ActivityFunction1 and ActivityFunction2 calls in another orchestration, and have the first orchestrator call that, would that help as well?

**Comment by amadard** [comment]
@davidychow87 No, the suborchestration wont fix the concurrent instance "rate limiter" issue, but I think the example above will do it well.

**Comment by davidychow87** [comment]
@amadard Thanks for the info

**Comment by davidychow87** [comment]
I will try the above example. This doesn't seem like it would mess with auto-scaling, right? For instance, if I scale based on CPU percentage, each VM that gets spun up will have it's own OrchestrationLimiter?

**Comment by ConnorMcMahon** [comment]
@davidychow87 

To answer your question, it depends on how you use entity ids. In the case Sebastian gave, he had a single global OrchestrationLimiter, as he only used a single deterministic key.

If you want to ensure that each VM can only process N orchestrations at a time, that is a different matter. In Durable Functions, for the most part we don't have a global notion of the number of VMs you are running on, so we need to use partitions as a proxy for this.

To illustrate matters, let's say that you want each VM to process N orchestrations at a time and you have a number P partitions.

If N is large and isn't required to be exact, I would just use the global orchestration limiter Sebastian provided, with a limit of P * N.  As we use a hashing approach, for large N, on average each partition will get approximately N orchestrations. As long as you have at least as many VMs as you have partitions, each VM will have at most one partition, and therefore will process roughly N concurrent orchestrations.

It sounds like in your case, you have a very low value for N (like 2). In that case, it will be more tricky to enforce with any precision. Again, I would recommend that you have at least as many VMs as partitions, so each VM has at most one partition.  The below modification to Sebastian's client code assumes you have a copy of `[our hashing class](https://github.com/Azure/durabletask/blob/master/src/DurableTask.AzureStorage/Fnv1aHashHelper.cs).

```csharp
    [FunctionName("Start")]
    public static async Task<IActionResult> Start(
        [HttpTrigger(AuthorizationLevel.Anonymous, "post")] HttpRequest req,
        [DurableClient] IDurableClient client)
    {
        string instanceId = Guid.NewGuid().ToString();
        uint partition = Fnv1aHashHelper.ComputeHash(instanceId);
        var entityId = new EntityId("limiter", partitionId.ToString());
        await client.SignalEntityAsync<IOrchestrationLimiter>(entityId, p => p.StartNewOrchestration(("ExampleOrchestration", "", instanceId)));
        return await client.WaitForCompletionOrCreateCheckStatusResponseAsync(req, instanceId, TimeSpan.FromSeconds(60));
    }    
}
```

**Comment by davidychow87** [comment]
Thanks for the info @ConnorMcMahon I will look into that as well

**Comment by absognety** [comment]
Is this solution a part of python durable extension library?

**Comment by sebastianburckhardt** [comment]
I am not sure about which durable entity features are supported in the Python API at this moment. Perhaps @davidmrdavid knows, I think he has done some work on this.

**Comment by davidmrdavid** [comment]
Hi @absognety, Durable Entities are now available in the Durable Functions for Python SDK, although their documentation is a bit sparse at the moment. IntelliSense should be able to help here though, as well as [this](https://github.com/Azure/azure-functions-durable-python/tree/dev/samples/counter_entity) example :)

All that being said, Connor's modified solution above makes use of our internal C# hashing class, which clearly is not something that would readily work in Python. That being said, I don't immediately see a reason why Sebastian's original limiter-entity shouldn't work. And if it doesn't due to some missing or faulty API, that's a bug on me and you can file an issue about it 😉 

**If y'all feel strongly that this is a feature you'd like to see,** could you please open a new ticket proposing this feature? We can use that thread to iterate and brainstorm on the right interface to specify this, as well as gauging user interest in this. I can't make any promises that we'll necessary tackle this, but we can at least start a conversation. _Thanks!_

**Comment by alacho22** [comment]
Just thought I would post here to save other .NET isolated worker adaptors some time. Unfortunately, [.NET isolated workers do not yet support Durable Entities](https://github.com/microsoft/durabletask-dotnet/issues/19) which means that @sebastianburckhardt 's solution can't be used. We have an upstream service which we don't control and would likely fall over if we send too many requests, therefore rate-limiting the number of in-flight orchestrations is essential in our instance. At the moment I think the only solution we have come across is to manually keep track of the number of running instances in a table as described in [this Stack Overflow answer](https://stackoverflow.com/a/64025143/10018957). However, it really does feel like something which should be able to be handled by Azure.
